// go-cqwebsocket/out/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CQEvent = exports.CQWebSocket = exports.CQTag = exports.CQ = exports.Interfaces = exports.Tags = void 0;
const CQWebsocket_1 = require("./CQWebsocket");
Object.defineProperty(exports, "CQEvent", { enumerable: true, get: function () { return CQWebsocket_1.CQEvent; } });
Object.defineProperty(exports, "CQWebSocket", { enumerable: true, get: function () { return CQWebsocket_1.CQWebSocket; } });
const Interfaces = __importStar(require("./Interfaces"));
exports.Interfaces = Interfaces;
const Tags = __importStar(require("./tags"));
exports.Tags = Tags;
const tags_1 = require("./tags");
Object.defineProperty(exports, "CQ", { enumerable: true, get: function () { return tags_1.CQ; } });
Object.defineProperty(exports, "CQTag", { enumerable: true, get: function () { return tags_1.CQTag; } });
